ambry-clustermap/src/main/java/com.github.ambry.clustermap/HelixClusterManager.java

```diff
@@ -180,16 +163,19 @@ public HelixClusterManager(ClusterMapConfig clusterMapConfig, String instanceNam
             manager.addLiveInstanceChangeListener(clusterChangeHandler);
             logger.info("Registered live instance change listeners for Helix manager at {}", zkConnectStr);
 
-            dcToRoutingTableSnapshotRef.computeIfAbsent(dcName, k -> new AtomicReference<>())
-                .getAndSet(routingTableProvider.getRoutingTableSnapshot());
-            // the initial routing table change should populate the instanceConfigs, If it's empty that means initial
+            // in case initial event occurs before adding routing table listener, here we explicitly set snapshot in
+            // ClusterChangeHandler. The reason is, if listener missed initial event, snapshot inside routing table
+            // provider should be already populated.
+            clusterChangeHandler.setRoutingTableSnapshot(routingTableProvider.getRoutingTableSnapshot());
+            // the initial routing table change should populate the instanceConfigs. If it's empty that means initial
             // change didn't come and thread should wait on the init latch to ensure routing table snapshot is non-empty
-            if (dcToRoutingTableSnapshotRef.get(dcName).get().getInstanceConfigs().size() == 0) {
+            if (dcToDcZkInfo.get(dcName).clusterChangeHandler.getRoutingTableSnapshot().getInstanceConfigs().size()
+                == 0) {
               // Periodic refresh in routing table provider is enabled by default. In worst case, routerUpdater should
-              // refresh snapshot and trigger routing table change within 5 minutes. So we make clusterChangeHandler wait
-              // here for initial notification until times out.
-              clusterChangeHandler.waitForInitNotification();
+              // trigger routing table change within 5 minutes
+              dcToDcZkInfo.get(dcName).clusterChangeHandler.waitForInitNotification();
```

Comment:
Add an info output? so that we know it run into a waiting.

Reply:
sure, will add